File: 2022/day07-2.hs
import Data.List (isSuffixOf, sort)
import Data.Text (intercalate, pack, split, splitOn, unpack)
import Debug.Trace (trace)

data Type = D | F deriving (Show, Eq)

type File = ([String], Int, Type)

type State = ([String], [File])

total :: Int
total = 70_000_000

required :: Int
required = 30_000_000

solution :: String -> Int
solution content = (!! 0) . sort . filter (>= toFree) . map (dirSize files) . filter isDirectory $ files
  where
    (_, files) = foldl parse ([], []) . lines $ content
    occupied = dirSize files (["/"], 0, D)
    toFree = required - total + occupied

dirSize :: [File] -> File -> Int
dirSize files (_, size, F) = size
dirSize files dir = sum . map fileSize . filter (isContainedIn dir) $ files

fileName :: File -> [String]
fileName = fstOf3

fileSize :: File -> Int
fileSize = sndOf3

isDirectory :: File -> Bool
isDirectory = (== D) . trdOf3

isContainedIn :: File -> File -> Bool
isContainedIn f1 f2 = fileName f1 `isSuffixOf` fileName f2

size :: [File] -> File -> Int
size files file = sum . map fileSize . filter (file `isContainedIn`) $ files

parse :: State -> String -> State
parse state ('$' : ' ' : 'l' : 's' : _) = state
parse (current, files) ('$' : ' ' : 'c' : 'd' : ' ' : '.' : '.' : _) = (tail current, files)
parse (current, files) ('$' : ' ' : 'c' : 'd' : ' ' : xs) = (xs : current, files)
parse (current, files) ('d' : 'i' : 'r' : ' ' : xs) = (current, (xs : current, 0, D) : files)
parse (current, files) line = (current, (name : current, read size, F) : files)
  where
    [size, name] = words line

fstOf3 :: (a, b, c) -> a
fstOf3 (a, _, _) = a

sndOf3 :: (a, b, c) -> b
sndOf3 (_, b, _) = b

trdOf3 :: (a, b, c) -> c
trdOf3 (_, _, c) = c

main :: IO ()
main = do readFile "day07.dat" >>= print . solution

File: 2022/day07_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any, Literal

type FileType = Literal["F"] | Literal["D"]
type FileName = list[str]
type FileSize = int

type File = tuple[FileName, FileSize, FileType]

type State = tuple[FileName, list[File]]


TOTAL = 70_000_000
REQUIRED = 30_000_000
ROOT = (["/"], 0, "D")


def solution(content: list[str]) -> Any:
    files = parse(content)
    occupied = directory_size(files, ROOT)
    to_free = REQUIRED - TOTAL + occupied
    return sorted(
        value
        for value in (
            directory_size(files, file) for file in files if is_directory(file)
        )
        if value >= to_free
    )[0]


def directory_size(files: list[File], file: File) -> FileSize:
    _, fsize, ftype = file
    if ftype == "F":
        return fsize
    return sum(file_size(item) for item in files if contains(file, item))


def is_directory(file: File) -> bool:
    _, _, tp = file
    return tp == "D"


def file_name(file: File) -> FileName:
    name, _, _ = file
    return name


def contains(f1: File, f2: File) -> bool:
    return "".join(file_name(f2)).endswith("".join(file_name(f1)))


def file_size(file: File) -> FileSize:
    _, fsize, _ = file
    return fsize


def size(files: list[File], file: File) -> FileSize:
    return sum(file_size(item) for item in files if contains(file, item))


def parse(content: list[str]) -> list[File]:
    state = ([], [])
    for line in content:
        state = parse_line(state, line)
    _, files = state
    return files


def parse_line(state: State, line: str) -> State:
    clean = line.strip()
    if clean.startswith("$ ls"):
        return state
    current, files = state
    if clean.startswith("$ cd .."):
        return (current[1:], files)
    if clean.startswith("$ cd "):
        return ([clean[5:], *current], files)
    if clean.startswith("dir "):
        return (current, [([clean[4:], *current], 0, "D"), *files])
    fsize, fname, *_ = clean.split(" ")
    return (current, [([fname, *current], int(fsize), "F"), *files])


def main() -> None:
    with open("day07.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
